# IMPORT ALL REQUIRED LIBRARIES AND LOAD DATASET

## Required Libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Load Dataset

In [4]:
nadp = pd.read_csv("C:\\Users\\saina\\Desktop\\DS_ML_AI\\Scaler\\Projects_or_Case_Studies_GIT\\Network_Anomaly_Detection_Project\\Datasets\\Network_anomaly_data.csv")

# BASIC EXPLORATORY DATA ANALYSIS

## First 5 rows

In [5]:
nadp.head()

duration protocoltype   service flag  srcbytes  dstbytes  land  \
0         0          tcp  ftp_data   SF       491         0     0   
1         0          udp     other   SF       146         0     0   
2         0          tcp   private   S0         0         0     0   
3         0          tcp      http   SF       232      8153     0   
4         0          tcp      http   SF       199       420     0   

   wrongfragment  urgent  hot  ...  dsthostsamesrvrate  dsthostdiffsrvrate  \
0              0       0    0  ...                0.17                0.03   
1              0       0    0  ...                0.00                0.60   
2              0       0    0  ...                0.10                0.05   
3              0       0    0  ...                1.00                0.00   
4              0       0    0  ...                1.00                0.00   

   dsthostsamesrcportrate  dsthostsrvdiffhostrate  dsthostserrorrate  \
0                    0.17                    0.00               0.00   
1                    0.88                    0.00               0.00   
2                    0.00                    0.00               1.00   
3                    0.03                    0.04               0.03   
4                    0.00                    0.00               0.00   

   dsthostsrvserrorrate  dsthostrerrorrate  dsthostsrvrerrorrate   attack  \
0                  0.00               0.05                  0.00   normal   
1                  0.00               0.00                  0.00   normal   
2                  1.00               0.00                  0.00  neptune   
3                  0.01               0.00                  0.01   normal   
4                  0.00               0.00                  0.00   normal   

   lastflag  
0        20  
1        15  
2        19  
3        21  
4        21  

[5 rows x 43 columns]

**Observations**
> 43 columns are present in the data

## Info

In [6]:
nadp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125973 entries, 0 to 125972
Data columns (total 43 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   duration                125973 non-null  int64  
 1   protocoltype            125973 non-null  object 
 2   service                 125973 non-null  object 
 3   flag                    125973 non-null  object 
 4   srcbytes                125973 non-null  int64  
 5   dstbytes                125973 non-null  int64  
 6   land                    125973 non-null  int64  
 7   wrongfragment           125973 non-null  int64  
 8   urgent                  125973 non-null  int64  
 9   hot                     125973 non-null  int64  
 10  numfailedlogins         125973 non-null  int64  
 11  loggedin                125973 non-null  int64  
 12  numcompromised          125973 non-null  int64  
 13  rootshell               125973 non-null  int64  
 14  suattempted         

**Observations**
> Out of 43 Columns, there are 15 float64 type columns, 24 int64 type columns, 4 object(String) type columns

> Dataset size is 41.3 MB

>   Total Number of rows are 125973.

>   It looks like there are no null values in any column. (Should check) 

## Null Value Detection

In [7]:
nadp.isna().sum()

duration                  0
protocoltype              0
service                   0
flag                      0
srcbytes                  0
dstbytes                  0
land                      0
wrongfragment             0
urgent                    0
hot                       0
numfailedlogins           0
loggedin                  0
numcompromised            0
rootshell                 0
suattempted               0
numroot                   0
numfilecreations          0
numshells                 0
numaccessfiles            0
numoutboundcmds           0
ishostlogin               0
isguestlogin              0
count                     0
srvcount                  0
serrorrate                0
srvserrorrate             0
rerrorrate                0
srvrerrorrate             0
samesrvrate               0
diffsrvrate               0
srvdiffhostrate           0
dsthostcount              0
dsthostsrvcount           0
dsthostsamesrvrate        0
dsthostdiffsrvrate        0
dsthostsamesrcportra

In [9]:
sum(nadp.isna().sum())

0

**Observations**
> There are no null values in the dataset. 

## Nunique of all columns

In [22]:
for i in nadp.columns:
    print(f"{i:<25} :   {nadp[i].nunique()}")

duration                  :   2981
protocoltype              :   3
service                   :   70
flag                      :   11
srcbytes                  :   3341
dstbytes                  :   9326
land                      :   2
wrongfragment             :   3
urgent                    :   4
hot                       :   28
numfailedlogins           :   6
loggedin                  :   2
numcompromised            :   88
rootshell                 :   2
suattempted               :   3
numroot                   :   82
numfilecreations          :   35
numshells                 :   3
numaccessfiles            :   10
numoutboundcmds           :   1
ishostlogin               :   2
isguestlogin              :   2
count                     :   512
srvcount                  :   509
serrorrate                :   89
srvserrorrate             :   86
rerrorrate                :   82
srvrerrorrate             :   62
samesrvrate               :   101
diffsrvrate               :   95
srvdiffhostra

**Observations**
> Most of the columns have int and float data types, so for most of the columns, the number of unique values are greater than 50

> Can check unique values for columns with nunique <= 50, to identify whether they are (continuous or discrete) and (nominal or ordinal)  

## Function to display all rows and columns

In [19]:
def display_all(df):     # For any Dataframe df
   with pd.option_context('display.max_rows',100): # Change number of rows accordingly
      with pd.option_context('display.max_columns',100): # Change number of columns accordingly
          display(df)

In [21]:
display_all(nadp.head())

duration protocoltype   service flag  srcbytes  dstbytes  land  \
0         0          tcp  ftp_data   SF       491         0     0   
1         0          udp     other   SF       146         0     0   
2         0          tcp   private   S0         0         0     0   
3         0          tcp      http   SF       232      8153     0   
4         0          tcp      http   SF       199       420     0   

   wrongfragment  urgent  hot  numfailedlogins  loggedin  numcompromised  \
0              0       0    0                0         0               0   
1              0       0    0                0         0               0   
2              0       0    0                0         0               0   
3              0       0    0                0         1               0   
4              0       0    0                0         1               0   

   rootshell  suattempted  numroot  numfilecreations  numshells  \
0          0            0        0                 0          0   
1          0            0        0                 0          0   
2          0            0        0                 0          0   
3          0            0        0                 0          0   
4          0            0        0                 0          0   

   numaccessfiles  numoutboundcmds  ishostlogin  isguestlogin  count  \
0               0                0            0             0      2   
1               0                0            0             0     13   
2               0                0            0             0    123   
3               0                0            0             0      5   
4               0                0            0             0     30   

   srvcount  serrorrate  srvserrorrate  rerrorrate  srvrerrorrate  \
0         2         0.0            0.0         0.0            0.0   
1         1         0.0            0.0         0.0            0.0   
2         6         1.0            1.0         0.0            0.0   
3         5         0.2            0.2         0.0            0.0   
4        32         0.0            0.0         0.0            0.0   

   samesrvrate  diffsrvrate  srvdiffhostrate  dsthostcount  dsthostsrvcount  \
0         1.00         0.00             0.00           150               25   
1         0.08         0.15             0.00           255                1   
2         0.05         0.07             0.00           255               26   
3         1.00         0.00             0.00            30              255   
4         1.00         0.00             0.09           255              255   

   dsthostsamesrvrate  dsthostdiffsrvrate  dsthostsamesrcportrate  \
0                0.17                0.03                    0.17   
1                0.00                0.60                    0.88   
2                0.10                0.05                    0.00   
3                1.00                0.00                    0.03   
4                1.00                0.00                    0.00   

   dsthostsrvdiffhostrate  dsthostserrorrate  dsthostsrvserrorrate  \
0                    0.00               0.00                  0.00   
1                    0.00               0.00                  0.00   
2                    0.00               1.00                  1.00   
3                    0.04               0.03                  0.01   
4                    0.00               0.00                  0.00   

   dsthostrerrorrate  dsthostsrvrerrorrate   attack  lastflag  
0               0.05                  0.00   normal        20  
1               0.00                  0.00   normal        15  
2               0.00                  0.00  neptune        19  
3               0.00                  0.01   normal        21  
4               0.00                  0.00   normal        21

**Observations**
> Use display_all function whenever there is a need to print all the columns or rows.

## Unique values of columns with nunique <= 50

In [27]:
for i in nadp.columns:
    if nadp[i].nunique() <= 50:
        print(f"{i}")
        print(f"{nadp[i].unique()}")
        print()

protocoltype
['tcp' 'udp' 'icmp']

flag
['SF' 'S0' 'REJ' 'RSTR' 'SH' 'RSTO' 'S1' 'RSTOS0' 'S3' 'S2' 'OTH']

land
[0 1]

wrongfragment
[0 3 1]

urgent
[0 1 3 2]

hot
[ 0  5  6  4  2  1 28 30 22 24 14  3 15 25 19 18 77 17 11  7 20 12  9 10
  8 21 33 44]

numfailedlogins
[0 2 1 3 4 5]

loggedin
[0 1]

rootshell
[0 1]

suattempted
[0 1 2]

numfilecreations
[ 0  1  8  4  2 15 13 29 19 18  6 14  5 21 17 40  3 20 11 38 23 10 27 25
 12 16 28 26  7  9 33 22 43 36 34]

numshells
[0 1 2]

numaccessfiles
[0 1 2 3 5 4 8 6 7 9]

numoutboundcmds
[0]

ishostlogin
[0 1]

isguestlogin
[0 1]

attack
['normal' 'neptune' 'warezclient' 'ipsweep' 'portsweep' 'teardrop' 'nmap'
 'satan' 'smurf' 'pod' 'back' 'guess_passwd' 'ftp_write' 'multihop'
 'rootkit' 'buffer_overflow' 'imap' 'warezmaster' 'phf' 'land'
 'loadmodule' 'spy' 'perl']

lastflag
[20 15 19 21 18 17 16 12 14 11  2 13 10  9  8  7  3  5  1  6  0  4]



**Observation**
> **SHOULD COMMENT HERE AFTER UNDERSTANDING ABOUT EVERY COLUMN** 

## Descriptive statistics

In [31]:
display_all(nadp.describe())

duration      srcbytes      dstbytes           land  wrongfragment  \
count  125973.00000  1.259730e+05  1.259730e+05  125973.000000  125973.000000   
mean      287.14465  4.556674e+04  1.977911e+04       0.000198       0.022687   
std      2604.51531  5.870331e+06  4.021269e+06       0.014086       0.253530   
min         0.00000  0.000000e+00  0.000000e+00       0.000000       0.000000   
25%         0.00000  0.000000e+00  0.000000e+00       0.000000       0.000000   
50%         0.00000  4.400000e+01  0.000000e+00       0.000000       0.000000   
75%         0.00000  2.760000e+02  5.160000e+02       0.000000       0.000000   
max     42908.00000  1.379964e+09  1.309937e+09       1.000000       3.000000   

              urgent            hot  numfailedlogins       loggedin  \
count  125973.000000  125973.000000    125973.000000  125973.000000   
mean        0.000111       0.204409         0.001222       0.395736   
std         0.014366       2.149968         0.045239       0.489010   
min         0.000000       0.000000         0.000000       0.000000   
25%         0.000000       0.000000         0.000000       0.000000   
50%         0.000000       0.000000         0.000000       0.000000   
75%         0.000000       0.000000         0.000000       1.000000   
max         3.000000      77.000000         5.000000       1.000000   

       numcompromised      rootshell    suattempted        numroot  \
count   125973.000000  125973.000000  125973.000000  125973.000000   
mean         0.279250       0.001342       0.001103       0.302192   
std         23.942042       0.036603       0.045154      24.399618   
min          0.000000       0.000000       0.000000       0.000000   
25%          0.000000       0.000000       0.000000       0.000000   
50%          0.000000       0.000000       0.000000       0.000000   
75%          0.000000       0.000000       0.000000       0.000000   
max       7479.000000       1.000000       2.000000    7468.000000   

       numfilecreations      numshells  numaccessfiles  numoutboundcmds  \
count     125973.000000  125973.000000   125973.000000         125973.0   
mean           0.012669       0.000413        0.004096              0.0   
std            0.483935       0.022181        0.099370              0.0   
min            0.000000       0.000000        0.000000              0.0   
25%            0.000000       0.000000        0.000000              0.0   
50%            0.000000       0.000000        0.000000              0.0   
75%            0.000000       0.000000        0.000000              0.0   
max           43.000000       2.000000        9.000000              0.0   

         ishostlogin   isguestlogin          count       srvcount  \
count  125973.000000  125973.000000  125973.000000  125973.000000   
mean        0.000008       0.009423      84.107555      27.737888   
std         0.002817       0.096612     114.508607      72.635840   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       2.000000       2.000000   
50%         0.000000       0.000000      14.000000       8.000000   
75%         0.000000       0.000000     143.000000      18.000000   
max         1.000000       1.000000     511.000000     511.000000   

          serrorrate  srvserrorrate     rerrorrate  srvrerrorrate  \
count  125973.000000  125973.000000  125973.000000  125973.000000   
mean        0.284485       0.282485       0.119958       0.121183   
std         0.446456       0.447022       0.320436       0.323647   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       1.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

         samesrvrate    diffsrvrate  srvdiffhostrate   dsthostcount  \
count  125973.00000

In [32]:
display_all(nadp.describe(include="object"))

protocoltype service    flag  attack
count        125973  125973  125973  125973
unique            3      70      11      23
top             tcp    http      SF  normal
freq         102689   40338   74945   67343

**Observations**
> For most of the columns have min, 25th, 50th, 75th percentile as zeros. Only the max value is present. So those columns will have the many outliers in box plot.

## Duplicates Check

In [36]:
nadp.duplicated().sum()

np.int64(0)